File: python/strings/suffix_array.py
def suffix_array(s):
    s += "$"
    n = len(s)
    p = sorted(range(n), key=lambda i: s[i])
    c = [0] * n
    for i in range(1, n):
        c[p[i]] = c[p[i - 1]] + (s[p[i]] != s[p[i - 1]])
    k = 0
    while (1 << k) < n:
        p.sort(key=lambda i: (c[i], c[(i + (1 << k)) % n]))
        nc = [0] * n
        for i in range(1, n):
            cur = (c[p[i]], c[(p[i] + (1 << k)) % n])
            prev = (c[p[i - 1]], c[(p[i - 1] + (1 << k)) % n])
            nc[p[i]] = nc[p[i - 1]] + (cur != prev)
        c = nc
        k += 1
    return p[1:]


def lcp_array(s, sa):
    n = len(s)
    rank = [0] * n
    for i, p in enumerate(sa):
        rank[p] = i
    h = 0
    lcp = [0] * (n - 1)
    for i in range(n):
        if rank[i] == n - 1:
            h = 0
            continue
        j = sa[rank[i] + 1]
        while i + h < n and j + h < n and s[i + h] == s[j + h]:
            h += 1
        lcp[rank[i]] = h
        if h:
            h -= 1
    return lcp

File: cpp/strings/suffix_array.cpp
#include <bits/stdc++.h>
using namespace std;
struct SuffixArray {
  string s;
  vector<int> sa, rank, lcp;
  SuffixArray() {}
  SuffixArray(string x) {
    build(x);
  }
  void build(string x) {
    s = x;
    int n = s.size();
    sa.resize(n);
    rank.resize(n);
    iota(sa.begin(), sa.end(), 0);
    for (int i = 0; i < n; i++)
      rank[i] = s[i];
    vector<int> tmp(n);
    for (int k = 1;; k <<= 1) {
      sort(sa.begin(), sa.end(), [&](int a, int b) {
        if (rank[a] != rank[b])
          return rank[a] < rank[b];
        int ra = a + k < n ? rank[a + k] : -1, rb = b + k < n ? rank[b + k] : -1;
        return ra < rb;
      });
      tmp[sa[0]] = 0;
      for (int i = 1; i < n; i++) {
        auto key = [&](int a) { return pair<int, int>{rank[a], a + k < n ? rank[a + k] : -1}; };
        tmp[sa[i]] = tmp[sa[i - 1]] + (key(sa[i - 1]) != key(sa[i]));
      }
      rank = tmp;
      if (rank[sa.back()] == n - 1)
        break;
    }
    lcp.assign(max(0, n - 1), 0);
    for (int i = 0, h = 0; i < n; i++) {
      int r = rank[i];
      if (r == n - 1) {
        h = 0;
        continue;
      }
      int j = sa[r + 1];
      while (i + h < n && j + h < n && s[i + h] == s[j + h])
        h++;
      lcp[r] = h;
      if (h)
        h--;
    }
  }
};
